code/modules/projectiles/guns/beam_rifle.dm

```diff
@@ -85,149 +180,173 @@
 
 /obj/item/weapon/gun/energy/beam_rifle/proc/update_slowdown()
-	if(scoped)
+	if(aiming)
 		slowdown = scoped_slow
 	else
 		slowdown = initial(slowdown)
 
 /obj/item/weapon/gun/energy/beam_rifle/Initialize()
 	. = ..()
 	START_PROCESSING(SSfastprocess, src)
+	zoom_speed_action = new(src)
+	zoom_lock_action = new(src)
 
 /obj/item/weapon/gun/energy/beam_rifle/Destroy()
 	STOP_PROCESSING(SSfastprocess, src)
+	set_user(null)
+	clear_tracer()
 	..()
 
-/obj/item/weapon/gun/energy/beam_rifle/zoom(user, forced_zoom)
-	. = ..()
-	scope(user, .)
-
-/obj/item/weapon/gun/energy/beam_rifle/proc/scope(mob/user, forced)
-	var/scoping
-	switch(forced)
-		if(TRUE)
-			scoping = TRUE
-		if(FALSE)
-			scoping = FALSE
-		else
-			scoping = !scoped
-	if(scoping)
-		spread = scoped_inaccuracy
-		recoil = scoped_recoil
-		scoped = TRUE
-		user << "<span class='boldnotice'>You bring your [src] up and use its scope...</span>"
-	else
-		spread = hipfire_inaccuracy
-		recoil = hipfire_recoil
-		scoped = FALSE
-		user << "<span class='boldnotice'>You lower your [src].</span>"
-	update_slowdown()
-
-/obj/item/weapon/gun/energy/beam_rifle/can_trigger_gun(var/mob/living/user)
-	if(!scoped && !noscope)
-		user << "<span class='userdanger'>This beam rifle can only be used while scoped!</span>"
-		return FALSE
-	. = ..(user)
-
 /obj/item/weapon/gun/energy/beam_rifle/emp_act(severity)
 	chambered = null
 	recharge_newshot()
 
-/obj/item/weapon/gun/energy/beam_rifle/proc/aiming_beam()
-	var/atom/A = current_user.client.mouseObject
-	if(!istype(A) || !A.loc)
-		return
-	var/turf/T = get_turf(current_user.client.mouseObject)
-	if(!istype(T))
+/obj/item/weapon/gun/energy/beam_rifle/proc/aiming_beam(force_update = FALSE)
+	var/diff = abs(aiming_lastangle - lastangle)
+	check_user()
+	if(diff < AIMING_BEAM_ANGLE_CHANGE_THRESHOLD && !force_update)
 		return
+	aiming_lastangle = lastangle
+	CHECK_TICK
 	var/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam/P = new
 	P.gun = src
 	P.wall_pierce_amount = wall_pierce_amount
 	P.structure_pierce_amount = structure_piercing
 	P.do_pierce = projectile_setting_pierce
-	P.preparePixelProjectile(current_user.client.mouseObject, T, current_user, current_user.client.mouseParams, 0)
 	if(aiming_time)
 		var/percent = ((100/aiming_time)*aiming_time_left)
 		P.color = rgb(255 * percent,255 * ((100 - percent) / 100),0)
 	else
 		P.color = rgb(0, 255, 0)
-	clear_tracers()
-	P.fire()
-
-/obj/item/weapon/gun/energy/beam_rifle/proc/clear_tracers()
-	for(var/I in current_tracers)
-		current_tracers -= I
-		var/obj/effect/temp_visual/projectile_beam/PB = I
-		qdel(PB)
+	var/turf/curloc = get_turf(src)
+	var/turf/targloc = get_turf(current_user.client.mouseObject)
+	if(!istype(targloc))
+		if(!istype(curloc))
+			return
+		targloc = get_turf_in_angle(lastangle, curloc, 10)
+	P.preparePixelProjectile(targloc, targloc, current_user, current_user.client.mouseParams, 0)
+	P.fire(lastangle)
+
+/obj/item/weapon/gun/energy/beam_rifle/proc/clear_tracer()
```

Comment:
Feels really wasteful having an extra proc just to call `QDEL_NULL(current_tracer)`. Either just use qdel null directly or make it a define.